File: src/prod/src/managed/Api/src/System/Fabric/Common/StartNodeDescription2.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Common
{
    using System;
    using System.Fabric.Interop;

    internal sealed class StartNodeDescription2
    {
        public StartNodeDescription2(
            StartNodeDescriptionKind descriptionKind,
            object value)
        {
            this.DescriptionKind = descriptionKind;
            this.Value = value;
        }

        public StartNodeDescriptionKind DescriptionKind
        {
            get;
            internal set;
        }

        public object Value
        {
            get;
            internal set;
        }

        internal IntPtr ToNative(PinCollection pinCollection)
        {
            var nativeStartNodeDescription2 = new NativeTypes.FABRIC_START_NODE_DESCRIPTION2();

            switch (this.DescriptionKind)
            {
                case StartNodeDescriptionKind.UsingNodeName:
                    nativeStartNodeDescription2.Kind = NativeTypes.FABRIC_START_NODE_DESCRIPTION_KIND.FABRIC_START_NODE_DESCRIPTION_KIND_USING_NODE_NAME;
                    nativeStartNodeDescription2.Value =
                        ((StartNodeDescriptionUsingNodeName)this.Value).ToNative(pinCollection);
                    break;
                case StartNodeDescriptionKind.UsingReplicaSelector:
                    nativeStartNodeDescription2.Kind = NativeTypes.FABRIC_START_NODE_DESCRIPTION_KIND.FABRIC_START_NODE_DESCRIPTION_KIND_USING_REPLICA_SELECTOR;
                    // not implemented now
                    nativeStartNodeDescription2.Value = IntPtr.Zero;
                    break;
                default:
                    nativeStartNodeDescription2.Kind = NativeTypes.FABRIC_START_NODE_DESCRIPTION_KIND.FABRIC_START_NODE_DESCRIPTION_KIND_INVALID;
                    nativeStartNodeDescription2.Value = IntPtr.Zero;
                    break;
            }

            return pinCollection.AddBlittable(nativeStartNodeDescription2);
        }

        internal static unsafe StartNodeDescription2 CreateFromNative(IntPtr nativeRaw)
        {
            NativeTypes.FABRIC_START_NODE_DESCRIPTION2 native = *(NativeTypes.FABRIC_START_NODE_DESCRIPTION2*)nativeRaw;
            StartNodeDescriptionUsingNodeName startNodeDescriptionUsingNodeName = StartNodeDescriptionUsingNodeName.CreateFromNative(native.Value);

            return new StartNodeDescription2(StartNodeDescriptionKind.UsingNodeName, startNodeDescriptionUsingNodeName);
        }
    }
}

File: src/prod/src/ServiceModel/management/FaultAnalysisService/StartNodeDescription2.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------


#include "stdafx.h"

using namespace Common;
using namespace Naming;
using namespace ServiceModel;
using namespace std;

using namespace Management::FaultAnalysisService;

StringLiteral const TraceComponent("StartNodeDescription2");

StartNodeDescription2::StartNodeDescription2()
: kind_(StartNodeDescriptionKind::None)
, value_()
{
}

StartNodeDescription2::StartNodeDescription2(StartNodeDescription2 const & other)
: kind_(other.Kind)
, value_(other.Value)
{
}

StartNodeDescription2::StartNodeDescription2(StartNodeDescription2 && other)
: kind_(move(other.kind_))
, value_(move(other.value_))
{
}

Common::ErrorCode StartNodeDescription2::FromPublicApi(
    FABRIC_START_NODE_DESCRIPTION2 const & publicDescription)
{
    switch (publicDescription.Kind)
    {
        case FABRIC_START_NODE_DESCRIPTION_KIND_USING_NODE_NAME:
            kind_ = StartNodeDescriptionKind::Enum::UsingNodeName;
            break;
        case FABRIC_START_NODE_DESCRIPTION_KIND_USING_REPLICA_SELECTOR:
            kind_ = StartNodeDescriptionKind::Enum::UsingReplicaSelector;
            break;
    }

    if (value_)
    {
        Trace.WriteWarning(TraceComponent, "before assignment value_ is null");
    }

    value_ = publicDescription.Value;

    if (value_)
    {
        Trace.WriteWarning(TraceComponent, "after assignment value_ is null");
    }

    return ErrorCodeValue::Success;
}

void StartNodeDescription2::ToPublicApi(__in Common::ScopedHeap & heap, __out FABRIC_START_NODE_DESCRIPTION2 & result) const
{
    switch (kind_)
    {
    case StartNodeDescriptionKind::Enum::UsingNodeName:
        result.Kind = FABRIC_START_NODE_DESCRIPTION_KIND_USING_NODE_NAME;
        break;
    case StartNodeDescriptionKind::Enum::UsingReplicaSelector:
        result.Kind = FABRIC_START_NODE_DESCRIPTION_KIND_USING_REPLICA_SELECTOR;
        break;
    }

    result.Value = value_;
}
